A bunch of internal quality-of-life changes to strumlines

dombomb64 opened the pull request:
<!-- Please read the Contributing Guide (https://github.com/FunkinCrew/Funkin/blob/main/docs/CONTRIBUTING.md) before submitting this PR. -->

<!-- Does this PR close any issues? If so, link them below. -->
## Linked Issues
None

<!-- Briefly describe the issue(s) fixed. -->
## Description
I decided to make a bunch of internal quality-of-life changes to the way strumlines work.

This PR adds the following things:
- Better support for modding in extra strumlines.
- The ability to disable any inputs you want for any strumline, or just the whole strumline. This can make the opponent miss (if the strumline's `canMiss` field is also set to `true`), which now actually plays miss sounds and animations if they have any.
- The ability to choose which characters a strumline affects.
- Various additions to `PlayState.instance.regenNoteData`.
  - The ability to load notes from another chart.
  - The ability to specifically target certain strumlines. If a player strumline is targeted, this updates `Highscore.tallies.totalNotes`.
  - The ability to force a certain side of the chart to be loaded onto the chosen strumlines.
  - The ability to partially load charts.
  - The ability to combine loaded note data with existing note data
  - The ability to offset loaded note data.
  - With these, it's possible to mix and match different sections of charts together.
- More similarities between the way notes are handled between strumlines.
- The ability to update a strumline's `isPlayer` field and just have it work.
- A new `hasVibrations` variable in each strumline for haptics code to check for instead of `isPlayer`.
- Various additions to `VocalGroup`.
  - The ability to add more voice groups (which are now the designated type `VocalGroupEntry`) and assign them to different characters.
  - The ability to prevent updating an entry's `volume` with the field `manualVolume`, which, if enabled, makes it so that volume must be changed per `FunkinSound`, which is useful for scripts that want to change the volume without interference from the base game.
- Proper use of the strumlines' `fadeInArrows` function, which had documentation saying that it was only supposed to be used during Freeplay.
- Implementation of an unused strumline function called `fadeOutArrows`, which was intended to be used at the end of a Freeplay song.
- A function for strumlines called `skipFadingArrows`, which instantly completes either of the two aforementioned tweens. Useful for scripts that set the individual positions of the arrows, which can also just run the correct fade function again for Freeplay after the positions are set (as long as they're intended to be static during the fade (and if not, then a custom tween is still possible)).

Planned features:
- (Basically canned because I could only come up with a few ideas, one of which was also just better as a normal function) ~~Some additional getter and setter functions that serve as shortcuts. (For example, a character's `strumlines` variable actually pointing to a function that compiles a list of all strumlines whose `characters` field contains that character.)~~
- (Canned because I have no idea how to add an extra controller and an additional multiplayer keyboard scheme, but I imagine it replacing the strumlines' `isPlayer` field with something like `playerNumber`, which would be an `Int` if controlled by a human, or `null` if controlled by a bot) ~~Internal support for P2 controls for the opponent. (If this wasn't the intent for the local multiplayer I heard of, please let me know.)~~

<!-- Include any relevant screenshots or videos. -->
## Screenshots/Videos

Showcasing a script that makes the player interrupt the opponent:

https://github.com/user-attachments/assets/580ebe72-788d-4f6d-aa22-1dd789ede10a

Interrupt script if anybody wants to try it out:
```haxe
import flixel.util.FlxTimer;
import funkin.modding.module.Module;
import funkin.play.PlayState;

class InterruptModule extends Module {
  var interruptTimer:FlxTimer;

  function new() {
    super("mario bros wii");
  }

  public function onNoteHit(event:HitNoteScriptEvent) {
    if (event.note.noteData.getMustHitNote()) {
      var disableInput = PlayState.instance.opponentStrumline.disableInput;
      var canMiss = PlayState.instance.opponentStrumline.canMiss;
      PlayState.instance.opponentStrumline.disableInput = true;
      PlayState.instance.opponentStrumline.canMiss = true;
      if (interruptTimer != null) interruptTimer.reset();
      else interruptTimer = new FlxTimer().start(1, function(_) {
        PlayState.instance.opponentStrumline.disableInput = disableInput;
        PlayState.instance.opponentStrumline.canMiss = canMiss;
      });
    }
  }
}
```

Lasercar committed:
Chart Editor playtest option for Results Screen

Lasercar committed:
Welcome back welcome music

amyspark-ng committed:
fix notelay sound when not in grid

Lasercar committed:
A big commit for a small fix
Cursor in chart editor breaks in square after placing and removing note

Lasercar committed:
Custom note kind field fix
Now gets hidden if the note selected isn't a custom one

MAJigsaw77 committed:
WIP Support for Newgrounds on Mobile.

MAJigsaw77 committed:
Dont shadow the NG `API` aswell as automaticaly close the webview.

MAJigsaw77 committed:
Polish the webview stuff.

dombomb64 committed:
Some cool strumline stuff

Adds:
- Better support for modding in extra strumlines.
- The ability to disable any inputs you want for any strumline, or just the whole strumline. This can make the opponent miss, which now actually plays miss animations if they have any.
- The ability to choose which characters a strumline affects.
- The ability to call PlayState.instance.regenNoteData and specifically target certain strums.
- More similarities between the way notes are handled between strumlines.

dombomb64 committed:
Some cool strumline stuff

Adds:
- Better support for modding in extra strumlines.
- The ability to disable any inputs you want for any strumline, or just the whole strumline. This can make the opponent miss, which now actually plays miss animations if they have any.
- The ability to choose which characters a strumline affects.
- The ability to call PlayState.instance.regenNoteData and specifically target certain strums.
- More similarities between the way notes are handled between strumlines.

dombomb64 committed:
Bro why was this edit there

dombomb64 committed:
Bro why was this edit there

sphis-sinco commented:
👍

dombomb64 committed:
Fix an issue with hold notes not getting dropped

dombomb64 committed:
Fix an issue with hold notes not getting dropped

dombomb64 committed:
Vibrations *should* be working...

dombomb64 committed:
Vibrations *should* be working...

dombomb64 committed:
Actually fix hold notes + Upgrade VoicesGroup

dombomb64 committed:
Actually fix hold notes + Upgrade VoicesGroup

dombomb64 committed:
Add even more features to regenNoteData

Also removed the isFirstCallSinceSongLoad parameter in favor of the generatedNoteData field.

dombomb64 committed:
Add even more features to regenNoteData

Also removed the isFirstCallSinceSongLoad parameter in favor of the generatedNoteData field.

dombomb64 committed:
Add a single "getter" because I ran out of ideas

dombomb64 committed:
Add a single "getter" because I ran out of ideas

dombomb64 committed:
Add another getter and setter idea + Upgrade arrow tweens

dombomb64 committed:
Add another getter and setter idea + Upgrade arrow tweens

dombomb64 committed:
Fix a NoteVibrationsHandler crash

dombomb64 committed:
Fix a NoteVibrationsHandler crash

dombomb64 committed:
Prevent bots from missing during a lag spike

dombomb64 committed:
Prevent bots from missing during a lag spike

dombomb64 committed:
Make more fields get set automatically

dombomb64 committed:
Make more fields get set automatically

dombomb64 committed:
Fix compatibility for stage switch scripts even though you really shouldn't load something mid-song

dombomb64 committed:
Fix compatibility for stage switch scripts even though you really shouldn't load something mid-song

dombomb64 committed:
Actually fix the stage thing

dombomb64 committed:
Actually fix the stage thing

dombomb64 committed:
I may be stupid

dombomb64 committed:
I may be stupid

Lasercar committed:
Chart - Fix notekinds changing on multi select

Also reverts a change by this author that caused selected events to be reset when the event toolbox is first opened.

dombomb64 committed:
Merge branch 'develop' into feature/strumline-stuff

dombomb64 committed:
Support hold notes overlapping other notes

dombomb64 committed:
Support hold notes overlapping other notes

Kade-github reviewed (approved):
idk abnormal it runs pretty damn well I think
<img width="1080" height="621" alt="image" src="https://github.com/user-attachments/assets/8ebd94a7-74a1-4946-8469-301c21711b60" />

will send it internally, get the mobile team and more testers. thanks !!

EliteMasterEric reviewed (changes requested):
Overall high potential (I'd love to see someone try to recreate that one song from Holiday mod with this) but a thorough code review reveals a bunch of odd choices that may leave technical debt to fix later.

Lasercar commented on source/funkin/play/character/BaseCharacter.hx:
```diff
@@ -517,15 +518,9 @@ class BaseCharacter extends Bopper
     if (event.eventCanceled) return;
 
-    if (event.note.noteData.getMustHitNote() && characterType == BF)
+    if (event.note.parentStrumline.characters.contains(this))
```
I'd have to redo my no animation note PR to account for these changes too, when this is merged.

dombomb64 replied:
Just made a commit that reintroduces `characterType` in the if checks. I also made an even easier way to cancel sing and miss animations when I noticed how there was no check for any sort of "`playAnim`" variable in these functions, which for some reason, turned out to already exist for ghost miss events.

Kade-github commented on source/funkin/play/PlayState.hx:
```diff
@@ -3154,7 +3346,10 @@ class PlayState extends MusicBeatSubState
     if (combo >= 10) comboPopUps.displayCombo(combo);
 
-    if (vocals != null) vocals.playerVolume = 1;
+    if (vocals != null) for (track in strumline.vocals)
```
Esp since it only triggers the vocals, though it's a very minor thing as you said.
https://github.com/dombomb64/Funkin/blob/48d284b3d2518d84af9d6b371a8c3b2c9426a651/source/funkin/play/PlayState.hx#L3349

dombomb64 replied:
Interesting... When I went to relocate the vocal code to where the function is called, I realized that it's already handled there. Huh.

Kade-github commented:
with eric's review, we'll wait for the comments I haven't marked as resolved to be addressed by @dombomb64.

dombomb64 commented:
Wow, that's a lot... I'll see what I can do.

FuroYT committed:
Change BGScrollingText to a single FlxText

BGScrollingText is now a single FlxText drawn multiple time in a single frame which is more optimized that a FlxSpriteGroup with multiple FlxSprites taking the text's graphic

dombomb64 committed:
Make characterType useful again + Add playAnim to note events

dombomb64 committed:
Make characterType useful again + Add playAnim to note events

dombomb64 committed:
Fix improper use of input system + Fix canMiss

dombomb64 committed:
Fix improper use of input system + Fix canMiss

dombomb64 commented on source/funkin/play/notes/Strumline.hx:
```diff
@@ -140,13 +146,80 @@ class Strumline extends FlxSpriteGroup
+          {
+            var direction:NoteDirection = strumlineNotes.members[i].direction;
+            var timestamp:haxe.Int64 = PreciseInputManager.getCurrentTimestamp();
+            PreciseInputManager.instance.onInputReleased.dispatch(
```
Yeah, I don't know what I was thinking here. That's been fixed to not use the input system.

dombomb64 committed:
Stage switcher compat REVOKED!!!!

dombomb64 committed:
Stage switcher compat REVOKED!!!!

dombomb64 committed:
Remove unnecessary vocal volume change

dombomb64 committed:
Remove unnecessary vocal volume change

dombomb64 committed:
DK Thumbs Up! *fanfare plays*

dombomb64 committed:
DK Thumbs Up! *fanfare plays*

dombomb64 commented on source/funkin/play/PlayState.hx:
```diff
@@ -2295,63 +2385,126 @@ class PlayState extends MusicBeatSubState
+     * @param clearNotes Whether this function will either add to or overwrite existing notes.
      */
-  function regenNoteData(startTime:Float = 0):Void
+  function regenNoteData(startTime:Float = 0, ?endTime:Float, timeOffset:Float = 0, ?chart:SongDifficulty, ?strumline:Strumline, ?strumlines:Array<Strumline>,
```
I'm so glad this change was well-received. I was so excited for it to be added when I first implemented it.

dombomb64 committed:
Oh how the tables have strongered

dombomb64 committed:
Oh how the tables have strongered

Hundrec commented:
You should fix up one more conflict before Eric and Kade take another look

dombomb64 committed:
Merge branch 'develop' into feature/strumline-stuff

FuroYT committed:
Added `funnyColor` backwards compatibility

dombomb64 commented:
It changes the assets submodule? I didn't even know or mean to do that. I'll try to fix it.

EliteMasterEric committed:
Adjust WinAPI to allow setting theme on specific windows.

Co-authored-by: MAJigsaw77 <MAJigsaw77@users.noreply.github.com>

PurSnake committed:
Chart Editor Character Previews

Co-Authored-By: ComedyLost <106041578+comedylost@users.noreply.github.com>

anysad committed:
Remove unused Chart Editor Context Menu code

KoloInDaCrib committed:
make volume settings optionally carry over to playtest

NotHyper-474 committed:
Fix a syntax error with the RegistryMacro and a partial fix for code completion

NotHyper-474 committed:
Add hardware keyboard detection on Android

NotHyper-474 committed:
Skip processing dead notes

NotHyper-474 committed:
fix: Fix incorrect NaN checks & simplify finite check

PurSnake committed:
Properly load variation scripts in the Chart Editor when playtesting

EliteMasterEric commented on source/funkin/play/notes/NoteVibrationsHandler.hx:
```diff
@@ -9,54 +9,44 @@ import funkin.util.HapticUtil.HapticsMode;
 class NoteVibrationsHandler
```
It's not necessarily a problem you have to solve anything for, I was just making a note of it.

EliteMasterEric reviewed (changes requested):
This is a PR I'd really like to merge soon, but I hadn't gotten around to a re-review yet. I added some additional comments to look at.

ninjamuffin99 committed:
more controls.hx cleanup, remove a few unused functions, and add `Void` return types to make checkstyle happier

ninjamuffin99 committed:
remove more unused vars in controls.hx

ninjamuffin99 committed:
remove keyboard "scheme" related code, completely unused

ninjamuffin99 committed:
remove the unused _P and _R stuff from the Action enum abstract

ninjamuffin99 committed:
update our custom FunkinAction check() function with newer logic from FlxAction

VirtuGuy committed:
Fix results screen scrolling text speed

FuroYT committed:
Preload the Result Screen Music loop is there is an intro

AbnormalPoof committed:
Fix the nametag erroneously showing as "Boyfriend" when entering character select as Pico

AbnormalPoof committed:
Get the first owned character ID and use that for results

charlesisfeline committed:
null checkin for set health icon

NotHyper-474 committed:
fix: Pause state not being properly set during cutscenes

VirtuGuy committed:
Fix anim editor onion skin offset problems

KarimAkra committed:
Fix stickers transition looking off-size when resizing the game during a song

EliteMasterEric committed:
Code cleanup and additional docs for WaveformSprite.

EliteMasterEric committed:
Implement a checkbox to hide the waveform, which improves FPS.

EliteMasterEric committed:
Pause the chart editor when focus is lost.

AbnormalPoof committed:
Re-export Pico's confirm backing card with BTA to bake the filters

Hundrec committed:
Don't play Stress Pico cutscene if charting

EliteMasterEric committed:
Add difficulty switches to the undo/redo history.

EliteMasterEric committed:
Switching difficulty now forces the selection to be cleared.

AbnormalPoof committed:
Update assets and add `getAtlasSettings()` for texture atlas characters

Co-authored-by: VirtuGuy <Wyatt2010shaw@gmail.com>

EliteMasterEric committed:
Fix a bug where cycling difficulties wasn't calculated properly when switching variations.

Hundrec committed:
Adjust lyrics for Monster and Winter Horrorland

Fix some errors and formatting

EliteMasterEric committed:
Implemented Clone Difficulty button in the Difficulty toolbox.

Fixes to Clone Difficulty dialog, and added Move Difficulty dialog.

assets submod

KoloInDaCrib committed:
fix starting a playtest opening pause menu

VirtuGuy committed:
Fix ghost tapping behavior when unpausing

moondroidcoder committed:
fix: Make lerping consistent across different FPS

AbnormalPoof committed:
Make Monster's `singLEFT-hold` offset consistent with `singLEFT`

AbnormalPoof committed:
Re-export THE CHUDDER!!!!

ninjamuffin99 committed:
use FlxMath.wrap() to make this cursor wrapping cleaner in CharSelectSubState

ninjamuffin99 committed:
clean up spamDirection code, use FlxDirectionFlags

ninjamuffin99 committed:
move character select cursor sprite code into it's own class

moondroidcoder committed:
Update lime ref

KoloInDaCrib committed:
better backwards compatibility for `FlxAtlasSprite` -> `FunkinSprite`

ninjamuffin99 committed:
early return for Chart Editor handleCursor()

AbnormalPoof committed:
Fix the cigarette not appearing in the Pico doppelganger cutscene

AbnormalPoof committed:
Fix the cursor lerping from the top left in character select

AbnormalPoof committed:
Adjust Darnell's idle animation

EliteMasterEric committed:
Update HXCPP to support building with C++20

MAJigsaw77 committed:
Remove `WINAPI_GetProcessMemoryWorkingSetSize` dublicate.

AbnormalPoof committed:
Revert "remove keyboard "scheme" related code, completely unused"

This reverts commit 7c8b7eee578b4c5caabf194d645b1473f233dd46.

AbnormalPoof committed:
Re-export Girlfriend to texture atlas

AbnormalPoof committed:
Re-export Boyfriend (Car) to texture atlas

MAJigsaw77 committed:
Improve library version checks.

AbnormalPoof committed:
Lower him a little

MAJigsaw77 committed:
Remove the yellow notice.

MAJigsaw77 committed:
Bump `hxvlc` to `2.2.5`.

AbnormalPoof committed:
Re-export the Parents in the Eggnog Erect cutscene with BTA

AbnormalPoof committed:
Re-export Pico's PERFECT rank animation with BTA

AbnormalPoof committed:
Re-export GF's character select sprites to remove a stray line

NotHyper-474 committed:
fix: Cursor not properly fading out when exiting CharSelect with backspace

EliteMasterEric committed:
Some optimization to highlighted note checks.

AbnormalPoof committed:
Add `getFramesWithKeyword()`

AbnormalPoof committed:
Bump `flixel-animate` and add `useRenderTexture` to `FunkinSprite` as an option.

NotHyper-474 committed:
fix: solve crash when deleting save data

AbnormalPoof committed:
Re-export the Freeplay difficulty stars using BTA

EliteMasterEric committed:
Clean up the Lime target config dropdown by removing unused targets like AIR and Flash.

EliteMasterEric committed:
Rewrite measure ticks to not get redrawn EVERY FRAME GODAM

ninjamuffin99 committed:
reduce code copypaste a tiny smidge for chart editor waveforms

ninjamuffin99 committed:
early return in playVocals() to remove a level of nesting

ninjamuffin99 committed:
fix char select inputs for keyb and controllers

ninjamuffin99 committed:
remove unused pitch stuff in CharSelect

EliteMasterEric committed:
Redo the backing color of the measure ticks.

EliteMasterEric committed:
Tweak vertical offset on the measure numbers.

ninjamuffin99 committed:
use UIStateMachine in FreeplayState instead of checking/modifying `controls.active` everywhere
thus making it easier to fix the "unlocked inputs" issue (fix FUNK-2021)

ninjamuffin99 committed:
change fpScoreDisplay to private variable

ninjamuffin99 committed:
change fp -> fpScoreDisplay, as a more descriptive variable name

moondroidcoder committed:
fix: Stop the game from muting in silent mode when you unfocus then focus

Co-Authored-By: Karim Akra <144803230+KarimAkra@users.noreply.github.com>

ninjamuffin99 committed:
remove all the letterSort.inputEnabled stuff, only use the FreeplayState's uiStateMachine to check if inputs should be enabled

ninjamuffin99 committed:
change capsuleOptionsMenu -> instSelectMenu in FreeplayState

ninjamuffin99 committed:
remove some redundant getControls() in FreeplayState, just access controls directly

ninjamuffin99 committed:
remove some redundant getControls() in FreeplayState, just access controls directly

AbnormalPoof committed:
Add global offsets for Pico and Nene's pixel variants

Lasercar committed:
Chart Editor - additional audio checks for vocal only songs

sector-a committed:
Don't add the current difficulty in `Difficulty` entry of the Pause Menu.

Revert "Don't add the current difficulty in `Difficulty` entry of the Pause Menu."

This reverts commit e0518f636dceeafa6dd00c922972afa5840e1161.

Do nothing if you change difficulty to the same one.

moondroidcoder committed:
Decrease blocksize compression for gfChill

VirtuGuy committed:
Fix difficulty toolbox showing wrong song name

AbnormalPoof committed:
Shift the flames how did this happen

moondroidcoder committed:
Exclude gf Texture Atlases from ASTC Compression

AbnormalPoof committed:
Re-export Tankman and Tankman (Bloody) to texture atlas

AbnormalPoof committed:
Make Tankman less shiny

AbnormalPoof committed:
Re-export `bf` and `bf-dark` to texture atlas form and sever the death animation from `bf-car`

AbnormalPoof committed:
De-sharpen some atlases and optimize Monster's atlas

AbnormalPoof committed:
Re-export Spooky Kids to remove a billion black lines

AbnormalPoof committed:
Re-export Girlfriend (Dark) to texture atlas

AbnormalPoof committed:
Re-export some Week 7 assets to texture atlases

MAJigsaw77 committed:
Implement `clearAssets` for `Android` when building an `.apk`.

KoloInDaCrib committed:
add frames to the event data toolbox

AbnormalPoof committed:
Re-export Pico (holding Nene) to texture atlas

AbnormalPoof committed:
Re-export Otis (Speaker Shooter) to texture atlas

MAJigsaw77 committed:
Use our fork of `newgrounds`.

Kade-github committed:
Crash Fix and Visual Bug Fix on InstSelect

moondroidcoder committed:
Remove old junk code

moondroidcoder committed:
fix: Add callbacks that fades in the pause button for video cutscenes

Hundrec committed:
Widen Events toolbox

KarimAkra committed:
Update lime ref to hide the ios home indicator

moondroidcoder committed:
Exclude pico holding nene from astc compression

AbnormalPoof committed:
Fine-tuned offsets for 2 characters

moondroidcoder committed:
Decrease blocksize for higher quality flashes on otis' gun bangs

AbnormalPoof committed:
Re-export the background Tankmen in Week 7 Erect as texture atlases

moondroidcoder committed:
Decrease blocksizes for rando and sniper in Week 7 Erect

Hundrec committed:
Remove Chart Editor tick trace and bump assets

AbnormalPoof committed:
Fix atlases moving to the wrong place on restart

Bundled with a ton of manual offset fixes and adjustments

AbnormalPoof committed:
Re-export BF to remove that one dot in the confirm animation

KarimAkra committed:
Fix the game docking to the bottom of the screen for iPads

EliteMasterEric committed:
Make the 'Art' submodule shallow by default

sector-a committed:
Prevent the week choosing by tapping when a week is selected

AbnormalPoof committed:
Bump `flixel-animate` and simplify rimlights for atlases

Hundrec committed:
Update art

EliteMasterEric committed:
Fix a crash when browsing for files while Pause on Unfocus is disabled.

EliteMasterEric committed:
Fix an issue where welcome theme would play during playtesting.

EliteMasterEric committed:
Clean up isPlaytesting check.

EliteMasterEric committed:
Fix an issue where welcome music would resume after focus lost while playtesting.

Hundrec committed:
Fix Chart Editor event values getting reset

Co-Authored-By: Kolo <67389779+koloindacrib@users.noreply.github.com>

EliteMasterEric committed:
Fix a bug where selecting multiple events would replace one with the other (plus cleaned up logging!)

AbnormalPoof committed:
Bump `polymod`

EliteMasterEric committed:
ANSIUtil.hx code style cleanup, plus use ANSI logging in registry classes!

EliteMasterEric committed:
Add ANSI logging to even MORE classes!

KarimAkra committed:
Remove the new pillarboxing system from mobile

Mobile is not resizable so we don't really need it there

AbnormalPoof committed:
Re-export Nene and Nene (Dark) to texture atlas

EliteMasterEric committed:
Fix measure ticks drawing beyond the end of the song.

AbnormalPoof committed:
Make multi-animateatlas characters use a `getAnimateAtlas` function instead of making a new `FunkinSprite`

MAJigsaw77 committed:
Fix `FlxMacro` from breaking the completion server.

dombomb64 committed:
Merge branch 'feature/strumline-stuff' of https://github.com/dombomb64/Funkin into feature/strumline-stuff

dombomb64 committed:
Merge branch 'FunkinCrew:main' into feature/strumline-stuff

dombomb64 committed:
Merge branch 'develop' into feature/strumline-stuff

dombomb64 committed:
Merge remote-tracking branch 'upstream/develop' into feature/strumline-stuff

AbnormalPoof committed:
Revert some bandaid fixes for animate atlas characters to make them appear properly

AbnormalPoof committed:
Set `cornerPosition` to 0

AbnormalPoof committed:
Fixes with the None character and Blazin' characters

AbnormalPoof committed:
Fix multiple characters

AbnormalPoof committed:
Fix spirit againnnnn

AbnormalPoof committed:
Take global offsets into account for the character preview

EliteMasterEric committed:
Added more new colors to the AnsiUtil

EliteMasterEric committed:
Fix a build issue caused by copy-pasting AnsiUtil to the tools package.

EliteMasterEric committed:
Disable "DEBUG"-level OpenFL messages to avoid logspam from shader stuff.

EliteMasterEric committed:
Add support for stacked patches, improve capabilities of test operations

NotHyper-474 committed:
Improve the math for the song name position

Hundrec committed:
Consistent Chart Editor Window vbox heights

MAJigsaw77 committed:
Bump `extension-admob`.

EliteMasterEric committed:
Fix a bug where opening the welcome screen would restart the theme music

AbnormalPoof committed:
Fix the rain shader not working at all

MightyTheArmiddilo committed:
Make space detect HaxeUI Focus in the Chart Editor

Kade-github committed:
Convert magic variable to Fixel Constant

ExtraCode75 committed:
Update OffsetMenu.hx

Kade-github committed:
add nene rexport commit

AbnormalPoof committed:
Bump submod to fix an issue with nene-tankmen

AbnormalPoof committed:
Fix Nene being too high in Pico's game over

AbnormalPoof committed:
Fix gf-dark not properly transitioning between lit and unlit

moondroidcoder committed:
Update flixel ref for FlxText fixes

EliteMasterEric committed:
Update Polymod to fix an issue with objects being casted to strings.

AbnormalPoof committed:
Revert `polymod` to fix everything

dombomb64 committed:
Merge branch 'feature/strumline-stuff' of https://github.com/dombomb64/Funkin into feature/strumline-stuff

dombomb64 committed:
Merge branch 'feature/strumline-stuff' of https://github.com/dombomb64/Funkin into feature/strumline-stuff

dombomb64 committed:
Merge commit '33cc0961d' into feature/strumline-stuff

dombomb64 committed:
Revert "Merge branch 'FunkinCrew:main' into feature/strumline-stuff"

This reverts commit 35616338a34fbbc067c8566d1467bf0aeecdf22d, reversing
changes made to 6c86bdd621a4b84ed8eff73de125ec2b9e71f224.

dombomb64 committed:
Some updating for Strumline QoL

dombomb64 committed:
Merge remote-tracking branch 'upstream/develop' into feature/strumline-stuff

anysad commented:
Please dombomb I need this... my strumlines are kinda are extraless...

dombomb64 committed:
Add strumline code from review

dombomb64 committed:
Merge remote-tracking branch 'upstream/develop' into feature/strumline-stuff

dombomb64 committed:
Fix some bugs involving hold notes and covers

Hundrec commented:
I'm going to close this PR as stale for now.
Feel free to open a new one based on the new `preview/public-playtest` branch if you'd like to!
